### PR TITLE
[sinks] Replace the transaction field with a Kafka header in the sink

### PR DESCRIPTION
**Pull request by bkirwi**
Aside from the code change, which is relatively small:
- For tests that filtered out the timestamp with regex, don't assert on it.
- For tests that asserted on the value of the old field, assert on the new header.
- Add support for extracting and asserting on headers in kafka-verify.
- Merge some duplicated code in kafka-verify to make the last point easier.

### Motivation

This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/14659

### Tips for reviewer

Test changes made this a bit of a largish PR, though most changes are pretty mechanical. I've commented on a few parts I'm less confident in.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Remove the `transaction` field from the Kafka sink schema.
  - Expose the timestamp of sunk messages as a new `mz-timestamp` header.

### CODE REVIEW

File: src/testdrive/src/action/kafka/verify.rs
Comment (bkirwi) [inline src/testdrive/src/action/kafka/verify.rs:192]: Possible weird cases here: missing header, duplicated header, non-string header, header with no value. The current code is pretty lenient, since I really have only one use-case for this so far... but we could always make it stricter now or in the future.

File: src/storage/src/sink/kafka.rs
Comment (bkirwi) [reply src/storage/src/sink/kafka.rs:1223]: OTOH [Confluent's blog uses dashes](https://www.confluent.io/blog/5-things-every-kafka-developer-should-know/#tip-5-record-headers). I love a good hyphen, so I'll leave it as-is unless I hear otherwise.

File: test/testdrive/kafka-json-sinks.td
Comment (cjubb39) [inline test/testdrive/kafka-json-sinks.td:61]: lol @ this coincidence that these happened to be 13 digits long

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:104]: nice!

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [reply src/storage/src/sink/kafka.rs:1223]: I have no real opinion but maybe you can drop a note in `#devex` on slack and someone might chime in?

But maybe we can make this a constant?

File: src/testdrive/src/action/kafka/verify.rs
Comment (cjubb39) [inline src/testdrive/src/action/kafka/verify.rs:202]: nit: `message.key().map(|bytes| bytes.to_owned())` or `message.key().map(ToOwned::to_owned)` are a little more idiomatic I think

**Review by cjubb39 — APPROVED** [review]
couple inline comments -- looks good!  only one that I think should definitely be addressed before merging is checking with devex about the naming of the header!

File: test/testdrive/kafka-json-sinks.td
Comment (bkirwi) [reply test/testdrive/kafka-json-sinks.td:61]: 😅

File: src/testdrive/src/action/kafka/verify.rs
Comment (bkirwi) [reply src/testdrive/src/action/kafka/verify.rs:202]: oof, yeah, not sure where i was going with this one...

File: test/kafka-multi-broker/03-after-join.td
Comment (bkirwi) [reply test/kafka-multi-broker/03-after-join.td:41]: Yeah, I think we're okay! We could assert that with the existing verify impl by using regex replacement like we did before. (As far as I could tell folks didn't actually care that the timestamps existed, and were just trying to get rid of a bit of nondeterminism.)

File: src/testdrive/src/action/kafka/verify.rs
Comment (bkirwi) [reply src/testdrive/src/action/kafka/verify.rs:202]: ah right, this is from the old implementation. might as well tidy up while we're here!

File: src/testdrive/src/action/kafka/verify.rs
Comment (bkirwi) [reply src/testdrive/src/action/kafka/verify.rs]: Cool - I'll crank this to be maximally strict then, and future users can relax it if they want. No point designing around use-cases we don't have yet...

File: src/testdrive/src/action/kafka/verify.rs
Comment (bkirwi) [reply src/testdrive/src/action/kafka/verify.rs:374]: Sure! Though if it actually catches any errors I may back it out and save it for another PR.

Comment (cjubb39) [reply src/testdrive/src/action/kafka/verify.rs:374]: makes sense!

looks like it worked out so great!

File: test/kafka-multi-broker/03-after-join.td
Comment (cjubb39) [reply test/kafka-multi-broker/03-after-join.td:41]: sounds good!

File: src/storage/src/sink/kafka.rs
Comment (cjubb39) [inline src/storage/src/sink/kafka.rs:1223]: looks like marta suggested on slack that we use `materialize` rather than `mz`

**Review by cjubb39 — APPROVED** [review]
will leave it to you to change the header name before merging based on slack convos if you like.  Otherwise, looks great!

**Review by cjubb39 — APPROVED** [review]
still lgtm!

**Pull request merged**
